Mling add a bunch of metrics

mollyling opened the pull request:
Separated attempts from successes/failures.  Then put in our failure rate instead of pure fail count.  Also counts on social networks.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2301)
<!-- Reviewable:end -->

mollyling commented:
Ping?  @dborkan, if you're too busy to take a look, lemme know and I'll reassign.  I'll be out for a few days, and I'd prefer that this doesn't sit around while I'm out.

dborkan commented:
I can review this, sorry for the delay

On Wed, Mar 16, 2016 at 1:32 PM, Molly Ling <notifications@github.com>
wrote:

> Ping? @dborkan <https://github.com/dborkan>, if you're too busy to take a
> look, lemme know and I'll reassign. I'll be out for a few days, and I'd
> prefer that this doesn't sit around while I'm out.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/uProxy/uproxy/pull/2301#issuecomment-197445452>
>

dborkan commented:
Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/generic_core/metrics.ts, line 53 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-jObm-IRaTBePtMqg:-KD-jObn-Zx25MxXy4Jn:1593593821) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L53)):*
I think updater is unused in this method

---

*[src/generic_core/metrics.ts, line 75 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-k2N95_MX0EcAkLdA:-KD-k2N95_MX0EcAkLdB:290830511) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L75)):*
I think you might be able to replace this with _.reduce from the lodash library.  That's already included in the uProxy, you could just require it in this file similar to how it's included in uproxy_core.ts

---

*[src/generic_core/metrics.ts, line 136 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-lOPaDNk76mYIcNTg:-KD-lOPaDNk76mYIcNTh:431247238) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L136)):*
can you comment on why you are doing 20 * 5.0, rather than just 100.0?

---

*[src/generic_core/metrics.ts, line 180 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KCbLBlge0yEWQrD2yZM:-KCbLBlge0yEWQrD2yZN:-1549990020) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L180)):*
I'm a bit confused how ```natMetric``` can work for both integers and strings.  Is it correct that we want to use natMetric for all these?

---

*[src/generic_core/metrics.ts, line 202 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-knqTMeCsvI16VuAO:-KD-knqTMeCsvI16VuAP:-1986806194) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L202)):*
For users who already have uProxy installed and have the old format of MetricsData written to there storage, is this code going to cause problems?  For example, storedData should still be an object, but ```storedData.on_facebook``` might be undefined

---

*[src/generic_core/metrics.ts, line 240 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-hvEHPFJjpENFDsJc:-KD-hvEIBML9betvOgSG:-215873390) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L240)):*
can you comment on why we have the max(...) expression?  Also just a heads up, it looks like we will be adding the ability to remove users from the roster in the near future (primarily for Cloud, since users may want to destroy servers they've deployed)

---

*[src/generic_core/metrics.ts, line 314 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-hJpK9h-3rSb5HusX:-KD-hJpK9h-3rSb5HusY:-832236682) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L314)):*
What happens for the Cloud and WeChat networks?  WeChat is just in the dev build and will probably remain there for a very long time, but Cloud is already on the dist build and in use by a few users and will be used by more soon.  Maybe we should just add metrics for those networks now?

---

*[src/generic_core/remote-connection.ts, line 326 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KCbMoTeoINfxkro0L4_:-KCbMoTf5iwyJjtEJZMZ:-637255814) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/remote-connection.ts#L326)):*
can we rename shutdown to stopGet or something like that?  "shutdown" sounds a bit too much like the user quit Chrome or logged out of uProxy.

---

*[src/generic_core/social.spec.ts, line 87 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-h1aW1HgQv2X9hqtm:-KD-h1aW1HgQv2X9hqtn:1264339229) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.spec.ts#L87)):*
do you need the ```this.``` here?  why not just ```metrics = new rappor_metrics.Metrics(null);``` since you've declared ```var metrics``` above?

---

*[src/generic_core/social.ts, line 181 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-fj4gcx0FU2HakYoE:-KD-fj4gcx0FU2HakYoF:1832646258) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.ts#L181)):*
I think you can use ```Object.keys(this.roster).length)``` rather than iterate through the object

---

*[src/generic_core/social.ts, line 187 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-gK0wWujNweZiE1bN:-KD-gK0xp8M-dT42Kl1o:-73995599) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.ts#L187)):*
Currently in network-options.ts we have a ```displayName``` field, maybe we should add a ```metricsName``` field to NETWORK_OPTIONS as well?  You can see ```areAllContactsUproxy``` in social.ts for an example of how to read NETWORK_OPTIONS

---

*[src/generic_core/social.ts, line 190 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-fntW-3f0MgX6fVIP:-KD-fntW-3f0MgX6fVIQ:1832503891) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.ts#L190)):*
Can you comment on why you are subtracting 1?  I think it's because you don't want to count the logged in user, but it would be good to clarify

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2301)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Review status: 0 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/generic_core/metrics.ts, line 53 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-jObm-IRaTBePtMqg:-KD-m94EJ8D7zDNh7_uB:-1073877315) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L53)):*
Yup, it's for the actual merge, if implemented later.

---

*[src/generic_core/metrics.ts, line 75 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-k2N95_MX0EcAkLdA:-KD-zGVnCoZc06-7gHvj:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L75)):*
Done.

---

*[src/generic_core/metrics.ts, line 136 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-lOPaDNk76mYIcNTg:-KD-yPq2NGJVB4SkII2q:-1384561783) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L136)):*
Renamed to calcQuantile, which should clear that up.

---

*[src/generic_core/metrics.ts, line 180 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KCbLBlge0yEWQrD2yZM:-KD-ysvdDJZjQ3n94juy:-1226883664) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L180)):*
Renamed to clarify.  Everything's a string at the rappor level.  The only difference is whether you use the raw input string or bucket them logarithmically.

---

*[src/generic_core/metrics.ts, line 202 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-knqTMeCsvI16VuAO:-KD-zZBZKGzJEstLSxmh:497492615) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L202)):*
Added a version check.  Good find, thanks.

---

*[src/generic_core/metrics.ts, line 240 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-hvEHPFJjpENFDsJc:-KD-zmXyvJ-vW3Dv5EvC:1419152853) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L240)):*
Great!  Thanks.  I've added a comment.

---

*[src/generic_core/metrics.ts, line 314 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-hJpK9h-3rSb5HusX:-KE89B-b0y3_RhcrKOth:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L314)):*
Done.

---

*[src/generic_core/remote-connection.ts, line 326 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KCbMoTeoINfxkro0L4_:-KE89LtN30qNEruCnwbH:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/remote-connection.ts#L326)):*
Done.

---

*[src/generic_core/social.spec.ts, line 87 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-h1aW1HgQv2X9hqtm:-KD05iZ9BugkXlm-THAD:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.spec.ts#L87)):*
Done.

---

*[src/generic_core/social.ts, line 181 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-fj4gcx0FU2HakYoE:-KD04fLeFi3RgEvq1PQg:-1834867277) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.ts#L181)):*
Thanks!

---

*[src/generic_core/social.ts, line 187 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-gK0wWujNweZiE1bN:-KE89PQVzP8Kxk5qdqUq:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.ts#L187)):*
Done.

---

*[src/generic_core/social.ts, line 190 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-fntW-3f0MgX6fVIP:-KD04kyTDNoePUuzvTag:367894369) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/social.ts#L190)):*
I moved the subtraction closer to the comment.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2301)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Thanks, please take another look.

---

Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-:-KE8AEu55RXkz9nYyCfg:-657065395)*
<!-- Sent from Reviewable.io -->

dborkan commented:
One really small comment, then :+1: - good stuff!!

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---

*[src/generic_core/metrics.ts, line 53 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-jObm-IRaTBePtMqg:-KEInOoUwLXhcn2pNMwD:-1024352054) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L53)):*
OK, want to add a comment on that, if you don't want this to be removed?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-:-KEIoFJGtZqlY41KDgF7:-1259127637)*
<!-- Sent from Reviewable.io -->

mollyling commented:
Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[src/generic_core/metrics.ts, line 53 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2301#-KD-jObm-IRaTBePtMqg:-KEX8A0pulSsgOo9c_-E:-499146009) ([raw file](https://github.com/uproxy/uproxy/blob/17e03ac4ac78ef88d7f23e10d863062a33fa5b46/src/generic_core/metrics.ts#L53)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2301)*
<!-- Sent from Reviewable.io -->
